src/main/java/uk/gov/hmcts/ccd/domain/service/casedataaccesscontrol/PseudoRoleToAccessProfileGenerator.java

```diff
@@ -0,0 +1,116 @@
+        Set<String> idamRoles = new HashSet<>();
+
+        idamRoles.addAll(getIdamRolesFromAcls(caseTypeDefinition.getAccessControlLists()));
+        idamRoles.addAll(getIdamRolesFromAcls(eventAcls(caseTypeDefinition)));
+        idamRoles.addAll(getIdamRolesFromAcls(stateAcls(caseTypeDefinition)));
+        idamRoles.addAll(getIdamRolesFromAcls(caseFieldAcls(caseTypeDefinition)));
+        idamRoles.addAll(getIdamRolesFromAcls(caseFieldComplexAcls(caseTypeDefinition)));
+
+        return idamRoles;
+    }
+
+    private Set<String> extractCaseRoles(CaseTypeDefinition caseTypeDefinition) {
+        Set<String> caseRoles = new HashSet<>();
+
+        caseRoles.add(CREATOR.getRole());
+        caseRoles.addAll(getCaseRolesFromAcls(caseTypeDefinition.getAccessControlLists()));
+        caseRoles.addAll(getCaseRolesFromAcls(eventAcls(caseTypeDefinition)));
+        caseRoles.addAll(getCaseRolesFromAcls(stateAcls(caseTypeDefinition)));
+        caseRoles.addAll(getCaseRolesFromAcls(caseFieldAcls(caseTypeDefinition)));
+        caseRoles.addAll(getCaseRolesFromAcls(caseFieldComplexAcls(caseTypeDefinition)));
+
+        return caseRoles;
+    }
+
+    private List<String> getCaseRolesFromAcls(List<AccessControlList> accessControlLists) {
+        return accessControlLists.stream()
+            .map(AccessControlList::getAccessProfile)
+            .filter(accessProfile -> accessProfile.matches(CASE_ROLE_ID_REGEX))
+            .collect(Collectors.toList());
+    }
+
+    private List<String> getIdamRolesFromAcls(List<AccessControlList> accessControlLists) {
+        return accessControlLists.stream()
+            .map(AccessControlList::getAccessProfile)
+            .filter(accessProfile -> !accessProfile.matches(CASE_ROLE_ID_REGEX))
+            .collect(Collectors.toList());
+    }
+
+    private List<AccessControlList> eventAcls(CaseTypeDefinition caseTypeDefinition) {
+        return caseTypeDefinition.getEvents().stream()
+            .flatMap(element -> element.getAccessControlLists().stream())
+            .collect(Collectors.toList());
+    }
+
+    private List<AccessControlList> stateAcls(CaseTypeDefinition caseTypeDefinition) {
+        return caseTypeDefinition.getStates().stream()
+            .flatMap(element -> element.getAccessControlLists().stream())
+            .collect(Collectors.toList());
+    }
+
+    private List<AccessControlList> caseFieldAcls(CaseTypeDefinition caseTypeDefinition) {
+        return caseTypeDefinition.getCaseFieldDefinitions().stream()
+            .flatMap(element -> element.getAccessControlLists().stream())
+            .collect(Collectors.toList());
+    }
+
+    private List<AccessControlList> caseFieldComplexAcls(CaseTypeDefinition caseTypeDefinition) {
+        return caseTypeDefinition.getCaseFieldDefinitions().stream()
+            .flatMap(element -> element.getComplexACLs().stream())
+            .collect(Collectors.toList());
+    }
```

Comment:
Some of the private methods in this class that just take in `CaseTypeDefinition` look like they could potentially be useful in future as public methods within `CaseTypeDefinition` instead?

Reply:
we can do that refactoring when it is needed then